### PR TITLE
Generic cache implementation

### LINKED ISSUE #485: Implement caching module
- choose some library or implement own cache
- reuse Bucket, Key and encodeKey from database module
- should support expiring

### PR DESCRIPTION
**Pull request by mpetrunic**
resolves #485

### CODE REVIEW

File: packages/lodestar/src/cache/abstract.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import {CacheId, CacheIdFunction} from "./types";
+import {clone, hashTreeRoot} from "@chainsafe/ssz";
+
+export abstract class CacheItem<T> {
+
+  protected sszType: AnySSZType;
+
+  protected idFunction: CacheIdFunction<T>;
+
+  protected cache: Map<CacheId, T> = new Map();
```
Comment (GregTheGreek) [inline packages/lodestar/src/cache/abstract.ts:15]: could we abstract this further? What if we wanted to use Redis ?

File: packages/lodestar/src/cache/abstract.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import {AnySSZType} from "@chainsafe/ssz-type-schema";
+import {CacheId, CacheIdFunction} from "./types";
+import {clone, hashTreeRoot} from "@chainsafe/ssz";
+
+export abstract class CacheItem<T> {
+
+  protected sszType: AnySSZType;
+
+  protected idFunction: CacheIdFunction<T>;
```
Comment (GregTheGreek) [inline packages/lodestar/src/cache/abstract.ts:13]: Comments explaining this?

File: packages/lodestar/src/cache/abstract.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  public clear(): void {
+    this.cache.clear();
+  }
+
+  public delete(id?: CacheId): void {
+    if(id) {
+      this.cache.delete(this.idToString(id));
+    } else {
```
Comment (GregTheGreek) [inline packages/lodestar/src/cache/abstract.ts]: I don't think we should have this - imo we should either mark `clear` private or don't allow for delete to call clear

File: packages/lodestar/src/cache/abstract.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  protected sszType: AnySSZType;
+
+  protected idFunction: CacheIdFunction<T>;
+
+  protected cache: Map<CacheId, T> = new Map();
+
+  protected constructor(sszType: AnySSZType, idFunction?: CacheIdFunction<T>) {
+    this.sszType = sszType;
+    this.idFunction = idFunction || this.hashRootIdFunction;
```
Comment (GregTheGreek) [inline packages/lodestar/src/cache/abstract.ts:25]: where does `this.hashRootIdFunction` come from? it will always be null?

**Review by GregTheGreek — DISMISSED** [review]

File: packages/lodestar/src/cache/abstract.ts (hunk 1 above)
Comment (mpetrunic) [reply packages/lodestar/src/cache/abstract.ts:15]: Can we really package redis with lodestar? I would leave further abstraction until we have some use case, should be fairly easy to abstract more

File: packages/lodestar/src/cache/abstract.ts (hunk 4 above)
Comment (mpetrunic) [reply packages/lodestar/src/cache/abstract.ts:25]: it's defined in class. Although I could move it to separate file with some common id generators

File: packages/lodestar/src/cache/abstract.ts (hunk 1 above)
Comment (GregTheGreek) [reply packages/lodestar/src/cache/abstract.ts:15]: no idea - just a thought

Comment (twoeths) [reply packages/lodestar/src/cache/abstract.ts:15]: Isn't it just installing redis npm package and then it's ready to use? It's in-memory so should be lightweight and doesn't require any special setup, it also supports expiring feature.

Comment (twoeths) [reply packages/lodestar/src/cache/abstract.ts:15]: if we support expiring feature then probably we'll end up with implement something like redis.

Comment (mpetrunic) [reply packages/lodestar/src/cache/abstract.ts:15]: I'm pretty sure thats just redis client. You need redis server for client to work. There are redis mocks, but they are a bit bloated and not maintained

**Comment by mpetrunic** [comment]
@ChainSafe/eth2-0 Can we review this?

**Comment by wemeetagain** [comment]
tbh i don't think we should go in this direction, rather explore using persistent immutable data structures.
This is because the object we're most likely to want to cache (state) doesn't change very much between blocks, and if we use a simple map of states, we will just be storing a lot of duplicate data.
This alternative path also fits nicely with the idea of persisting the merkle tree.

EDIT:
On second thought, we could still use something like this despite using different data structures.
I think my issue with this is I'd prefer to start with something simple and extensible, and grow something from there, and this feels like a full-blown caching framework.

**Comment by mpetrunic** [comment]
> tbh i don't think we should go in this direction, rather explore using persistent immutable data structures.
> This is because the object we're most likely to want to cache (state) doesn't change very much between blocks, and if we use a simple map of states, we will just be storing a lot of duplicate data.
> This alternative path also fits nicely with the idea of persisting the merkle tree.
> 
> EDIT:
> On second thought, we could still use something like this despite using different data structures.
> I think my issue with this is I'd prefer to start with something simple and extensible, and grow something from there, and this feels like a full-blown caching framework.

I was thinking of using this for api for now, and expand later for storing states and stuff. Like caching ValidatorDuties and commitees so validators don't block us on start of every slot.

Imean it's simple to extend for storing parts of state for example, it's more about interfaces (there is not much implementation anyways)

**Pull request closed without merging**